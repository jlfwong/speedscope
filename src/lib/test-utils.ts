import * as fs from 'fs'
import * as path from 'path'
import {Profile, CallTreeNode, Frame} from './profile'
import {importProfileGroup} from '../import'
import {exportProfileGroup, importSpeedscopeProfiles} from './file-format'

interface DumpedProfile {
  name: string
  stacks: string[]
  frames: Frame[]
}

export function dumpProfile(profile: Profile): any {
  const dump: DumpedProfile = {
    name: profile.getName(),
    stacks: [],
    frames: [],
  }

  profile.forEachFrame(f => dump.frames.push(f))

  let lastValue = 0
  const curStack: (number | string)[] = []

  function maybeEmit(value: number) {
    if (lastValue != value) {
      dump.stacks.push(
        curStack.map(k => `${k}`).join(';') + ` ${profile.formatValue(value - lastValue)}`,
      )
      lastValue = value
    }
  }

  function openFrame(node: CallTreeNode, value: number) {
    maybeEmit(value)
    curStack.push(node.frame.name)
  }

  function closeFrame(node: CallTreeNode, value: number) {
    maybeEmit(value)
    curStack.pop()
  }

  profile.forEachCall(openFrame, closeFrame)

  return dump
}

export async function checkProfileSnapshot(filepath: string) {
  const input = fs.readFileSync(filepath, 'utf8')

  const profileGroup = await importProfileGroup(path.basename(filepath), input)
  if (profileGroup) {
    expect(profileGroup.name).toMatchSnapshot('profileGroup.name')
    expect(profileGroup.indexToView).toMatchSnapshot('indexToView')

    for (let profile of profileGroup.profiles) {
      expect(dumpProfile(profile)).toMatchSnapshot()
    }
  } else {
    fail('Failed to extract profile')
    return
  }

  const profilesWithoutFilename = await importProfileGroup('unknown', input)
  if (profilesWithoutFilename) {
    expect(profilesWithoutFilename.profiles.length).toEqual(profileGroup.profiles.length)
    profilesWithoutFilename.name = profileGroup.name
    for (let i = 0; i < profileGroup.profiles.length; i++) {
      const a = profileGroup.profiles[i]
      const b = profilesWithoutFilename.profiles[i]
      b.setName(a.getName())
    }
    expect(exportProfileGroup(profileGroup)).toEqual(exportProfileGroup(profilesWithoutFilename))
  } else {
    fail('Failed to extract profile when filename was "unknown"')
    return
  }

  const exported = exportProfileGroup(profileGroup)
  const reimportedGroup = importSpeedscopeProfiles(exported)

  expect(reimportedGroup.name).toEqual(profileGroup.name)
  expect(reimportedGroup.profiles.length).toEqual(profileGroup.profiles.length)

  for (let i = 0; i < profileGroup.profiles.length; i++) {
    const profile = profileGroup.profiles[i]
    const reimported = reimportedGroup.profiles[i]

    expect(reimported.getTotalWeight()).toEqual(profile.getTotalWeight())
    expect(dumpProfile(reimported).stacks.join('\n')).toEqual(
      dumpProfile(profile).stacks.join('\n'),
    )
  }

  const reexported = exportProfileGroup(reimportedGroup)
  expect(exported).toEqual(reexported)
}
