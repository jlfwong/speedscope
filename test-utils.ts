import * as fs from 'fs'
import * as path from 'path'
import {Profile, CallTreeNode, Frame} from './profile'
import {importProfile} from './import'

interface DumpedProfile {
  stacks: string[]
  frames: Frame[]
}

export function dumpProfile(profile: Profile): any {
  const dump: DumpedProfile = {
    stacks: [],
    frames: [],
  }

  profile.forEachFrame(f => dump.frames.push(f))

  let lastValue = 0
  const curStack: (number | string)[] = []

  function maybeEmit(value: number) {
    if (lastValue != value) {
      dump.stacks.push(
        curStack.map(k => `${k}`).join(';') + ` ${profile.formatValue(value - lastValue)}`,
      )
      lastValue = value
    }
  }

  function openFrame(node: CallTreeNode, value: number) {
    maybeEmit(value)
    curStack.push(node.frame.name)
  }

  function closeFrame(node: CallTreeNode, value: number) {
    maybeEmit(value)
    curStack.pop()
  }

  profile.forEachCall(openFrame, closeFrame)

  return dump
}

export async function checkProfileSnapshot(filepath: string) {
  const input = fs.readFileSync(filepath, 'utf8')
  const profile = await importProfile(path.basename(filepath), input)
  if (profile) {
    expect(dumpProfile(profile)).toMatchSnapshot()
  } else {
    fail('Failed to extract profile')
  }
}
