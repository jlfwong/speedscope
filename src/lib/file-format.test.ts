import {checkProfileSnapshot} from './test-utils'

describe('importSpeedscopeProfiles', async () => {
  test('0.0.1 evented profile', async () => {
    await checkProfileSnapshot('./sample/profiles/speedscope/0.0.1/simple.speedscope.json')
  })

  test('0.1.2 sampled profile', async () => {
    await checkProfileSnapshot('./sample/profiles/speedscope/0.1.2/simple-sampled.speedscope.json')
  })

  test('0.6.0 multiple profiles', async () => {
    await checkProfileSnapshot('./sample/profiles/speedscope/0.6.0/two-sampled.speedscope.json')
  })
})
