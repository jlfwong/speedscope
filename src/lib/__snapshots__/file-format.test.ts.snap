// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`importSpeedscopeProfiles 0.0.1 evented profile 1`] = `
Object {
  "frames": Array [
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 0,
      "line": undefined,
      "name": "a",
      "selfWeight": 0,
      "totalWeight": 14,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 1,
      "line": undefined,
      "name": "b",
      "selfWeight": 5,
      "totalWeight": 14,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 2,
      "line": undefined,
      "name": "c",
      "selfWeight": 5,
      "totalWeight": 5,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 3,
      "line": undefined,
      "name": "d",
      "selfWeight": 4,
      "totalWeight": 4,
    },
  ],
  "name": "simple.txt",
  "stacks": Array [
    "a;b;c 2",
    "a;b;d 4",
    "a;b;c 3",
    "a;b 5",
  ],
}
`;

exports[`importSpeedscopeProfiles 0.0.1 evented profile: indexToView 1`] = `0`;

exports[`importSpeedscopeProfiles 0.0.1 evented profile: profileGroup.name 1`] = `"simple.txt"`;

exports[`importSpeedscopeProfiles 0.1.2 sampled profile 1`] = `
Object {
  "frames": Array [
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 0,
      "line": undefined,
      "name": "a",
      "selfWeight": 0,
      "totalWeight": 14,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 1,
      "line": undefined,
      "name": "b",
      "selfWeight": 5,
      "totalWeight": 14,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 2,
      "line": undefined,
      "name": "c",
      "selfWeight": 5,
      "totalWeight": 5,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 3,
      "line": undefined,
      "name": "d",
      "selfWeight": 4,
      "totalWeight": 4,
    },
  ],
  "name": "simple.speedscope.json",
  "stacks": Array [
    "a;b;c 2.00s",
    "a;b;d 4.00s",
    "a;b;c 3.00s",
    "a;b 5.00s",
  ],
}
`;

exports[`importSpeedscopeProfiles 0.1.2 sampled profile: indexToView 1`] = `0`;

exports[`importSpeedscopeProfiles 0.1.2 sampled profile: profileGroup.name 1`] = `"simple.speedscope.json"`;

exports[`importSpeedscopeProfiles 0.6.0 multiple profiles 1`] = `
Object {
  "frames": Array [
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 0,
      "line": undefined,
      "name": "a",
      "selfWeight": 0,
      "totalWeight": 14,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 1,
      "line": undefined,
      "name": "b",
      "selfWeight": 5,
      "totalWeight": 14,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 2,
      "line": undefined,
      "name": "c",
      "selfWeight": 5,
      "totalWeight": 5,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 3,
      "line": undefined,
      "name": "d",
      "selfWeight": 4,
      "totalWeight": 4,
    },
  ],
  "name": "one",
  "stacks": Array [
    "a;b;c 2.00s",
    "a;b;d 4.00s",
    "a;b;c 3.00s",
    "a;b 5.00s",
  ],
}
`;

exports[`importSpeedscopeProfiles 0.6.0 multiple profiles 2`] = `
Object {
  "frames": Array [
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 0,
      "line": undefined,
      "name": "a",
      "selfWeight": 0,
      "totalWeight": 14,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 1,
      "line": undefined,
      "name": "b",
      "selfWeight": 5,
      "totalWeight": 14,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 2,
      "line": undefined,
      "name": "c",
      "selfWeight": 5,
      "totalWeight": 5,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 3,
      "line": undefined,
      "name": "d",
      "selfWeight": 4,
      "totalWeight": 4,
    },
  ],
  "name": "two",
  "stacks": Array [
    "a;b;c 2.00s",
    "a;b;d 4.00s",
    "a;b;c 3.00s",
    "a;b 5.00s",
  ],
}
`;

exports[`importSpeedscopeProfiles 0.6.0 multiple profiles: indexToView 1`] = `1`;

exports[`importSpeedscopeProfiles 0.6.0 multiple profiles: profileGroup.name 1`] = `"Two Samples"`;
